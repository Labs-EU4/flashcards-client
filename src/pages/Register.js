import React, {useState} from "react";
import {connect} from "react-redux";
import {registerNewUser} from "../state/userData/userDataActionCreators";
import {useHistory} from "react-router-dom";
import {Form, Input, Button, Icon} from "antd";
import styles from "./Register.module.css";

export function RegisterForm({registerNewUser, ...props}) {
  let history = useHistory();
  const [emailInfo, setEmailInfo] = useState({emailValidationStatus: null, help: null});
  const [usernameInfo, setUsernameInfo] = useState({
    usernameValidationStatus: null,
    help: null,
  });
  const [passwordInfo, setPasswordInfo] = useState({
    passwordValidationStatus: null,
    help: null,
  });
  const [registerDisabled, setRegisterDisabled] = useState(false);
  const [isLoading, setIsLoading] = useState(false);

  if (localStorage.getItem("token")) {
    history.push("/");
  }

  const defaultInputs = {
    email: "",
    fullName: "",
    password: "",
  };
  const [user, setUser] = useState(defaultInputs);

  const handleChange = e => {
    setUser({
      ...user,
      [e.target.name]: e.target.value,
    });
  };

  async function handleSubmit(e) {
    e.preventDefault();
    setIsLoading(true);
    const {email, fullName, password} = e.target;
    try {
      if (email.value && fullName.value && password.value) {
        setRegisterDisabled(true);
        await registerNewUser(user);
        setIsLoading(false);
        setUser(defaultInputs);
        setRegisterDisabled(false);
      } else if (!email.value) {
        setEmailInfo({emailValidationStatus: "error", help: "Please enter an Email."});
        setRegisterDisabled(true);
      } else if (!fullName.value) {
        setUsernameInfo({
          usernameValidationStatus: "error",
          help: "Please enter a Username.",
        });
        setRegisterDisabled(true);
      } else if (!password.value) {
        setPasswordInfo({
          passwordValidationStatus: "error",
          help: "Please enter a Password.",
        });
        setRegisterDisabled(true);
      }
    } catch (err) {
      setIsLoading(false);
      setRegisterDisabled(false);
      console.error(err.response.data.message);
    }
  }

  function emailValidation(e) {
    const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    let inputString = e.target.value;

    const email = inputString.match(emailRegex);

    if (email) {
      setEmailInfo({emailValidationStatus: "success", help: null});
      setRegisterDisabled(false);
    } else {
      setEmailInfo({emailValidationStatus: "warning", help: "Not a valid email"});
      setRegisterDisabled(true);
    }
  }

  function usernameValidation(e) {
    let inputString = e.target.value;

    if (inputString.length >= 5) {
      setUsernameInfo({usernameValidationStatus: "success", help: null});
      setRegisterDisabled(false);
    } else {
      setUsernameInfo({
        usernameValidationStatus: "warning",
        help: "Username must be at least 5 characters",
      });
      setRegisterDisabled(true);
    }
  }

  function passwordValidation(e) {
    let inputString = e.target.value;

    if (inputString.length >= 5) {
      setPasswordInfo({passwordValidationStatus: "success", help: null});
      setRegisterDisabled(false);
    } else {
      setPasswordInfo({
        passwordValidationStatus: "warning",
        help: "Password must be at least 5 characters",
      });
      setRegisterDisabled(true);
    }
  }

  return (
    <div className={styles.registerContainer}>
      <Form onSubmit={event => handleSubmit(event)} className={styles.registerForm}>
        <Form.Item
          hasFeedback
          validateStatus={emailInfo.emailValidationStatus}
          help={emailInfo.help}
        >
          <Input
            onBlur={e => emailValidation(e)}
            prefix={<Icon type="mail" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="email"
            type="text"
            name="email"
            setfieldvalue={user.email}
            onChange={event => handleChange(event)}
          />
        </Form.Item>
        <Form.Item
          hasFeedback
          validateStatus={usernameInfo.usernameValidationStatus}
          help={usernameInfo.help}
        >
          <Input
            onBlur={e => usernameValidation(e)}
            prefix={<Icon type="user" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="username"
            type="text"
            name="fullName"
            setfieldvalue={user.fullName}
            onChange={event => handleChange(event)}
          />
        </Form.Item>
        <Form.Item
          hasFeedback
          validateStatus={passwordInfo.passwordValidationStatus}
          help={passwordInfo.help}
        >
          <Input
            onBlur={e => passwordValidation(e)}
            prefix={<Icon type="lock" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="password"
            type="password"
            name="password"
            setfieldvalue={user.password}
            onChange={event => handleChange(event)}
          />
        </Form.Item>
        <Form.Item>
          <Button
            type="primary"
            htmlType="submit"
            className={styles.registerFormButton}
            disabled={registerDisabled}
            loading={isLoading}
          >
            Register
          </Button>
          Or <a href="/login">Login Here!</a>
        </Form.Item>
      </Form>
    </div>
  );
}

function mapStateToProps(state) {
  return {
    userData: state.userData,
  };
}

const ConnectedForm = connect(mapStateToProps, {registerNewUser})(RegisterForm);

export default Form.create()(ConnectedForm);
