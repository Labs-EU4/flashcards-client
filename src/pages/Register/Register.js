import React, {useState} from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import {Form, Input, Button, Icon, Alert} from "antd";
import {registerNewUser} from "../../state/actions/auth";
import {baseURL} from "../../utils/axios";
import styles from "./Register.module.css";
import backgroundStyles from "../../components/formStyleComponent/FormStyleComponent.module.css";
import FormHeader from "../../components/formStyleComponent/FormHeader";

export function RegisterForm({registerNewUser, ...props}) {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState("");

  const [formInfo, setFormInfo] = useState({
    email: {validationStatus: null, help: null},
    username: {validationStatus: null, help: null},
    password: {validationStatus: null, help: null},
  });

  const [user, setUser] = useState({
    email: "",
    fullName: "",
    password: "",
  });

  const handleChange = e => {
    setUser({
      ...user,
      [e.target.name]: e.target.value,
    });
  };

  const handleSubmit = async e => {
    e.preventDefault();

    const {email, fullName, password} = user;

    try {
      if (email && fullName && password) {
        setIsLoading(true);
        await registerNewUser(user);
      } else {
        if (!email) {
          setFormInfo({
            ...formInfo,
            email: {validationStatus: "error", help: "Please enter an Email"},
          });
        }
        if (!fullName) {
          setFormInfo({
            ...formInfo,
            username: {validationStatus: "error", help: "Please enter a Username"},
          });
        }
        if (!password) {
          setFormInfo({
            ...formInfo,
            password: {validationStatus: "error", help: "Please enter a Password"},
          });
        }
      }
    } catch (error) {
      error.response
        ? setError(error.response.data.message)
        : setError("Something went wrong!");
    } finally {
      setIsLoading(false);
      setUser({
        email: "",
        fullName: "",
        password: "",
      });
      setFormInfo({
        email: {validationStatus: null, help: null},
        username: {validationStatus: null, help: null},
        password: {validationStatus: null, help: null},
      });
    }
  };

  function formValidation(e) {
    let inputString = e.target.value;
    let inputType = e.target.name;
    if (inputType === "email") {
      const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
      const email = inputString.match(emailRegex);
      if (email) {
        setFormInfo({...formInfo, email: {validationStatus: "success", help: null}});
      } else {
        setFormInfo({
          ...formInfo,
          email: {validationStatus: "warning", help: "Not a valid email"},
        });
      }
    } else if (inputType === "fullName") {
      if (inputString.length >= 5) {
        setFormInfo({...formInfo, username: {validationStatus: "success", help: null}});
      } else {
        setFormInfo({
          ...formInfo,
          username: {
            validationStatus: "warning",
            help: "Username must be at least 5 characters",
          },
        });
      }
    } else if (inputType === "password") {
      if (inputString.length >= 5) {
        setFormInfo({...formInfo, password: {validationStatus: "success", help: null}});
      } else {
        setFormInfo({
          ...formInfo,
          password: {
            validationStatus: "warning",
            help: "Password must be at least 5 characters",
          },
        });
      }
    }
  }

  return (
    <div className={backgroundStyles.formStyle} data-testid="test_register_container">
      <FormHeader />
      <h1>Sign-up.</h1>
      <Form
        onSubmit={handleSubmit}
        className={styles.registerForm}
        data-testid="test_register_form"
      >
        <Form.Item
          data-testid="test_email_form_item"
          hasFeedback
          validateStatus={formInfo.email.validationStatus}
          help={formInfo.email.help}
        >
          <Input
            data-testid="test_email_input"
            onBlur={e => formValidation(e)}
            prefix={<Icon type="mail" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="email"
            type="text"
            name="email"
            value={user.email}
            onChange={event => handleChange(event)}
          />
        </Form.Item>
        <Form.Item
          data-testid="test_username_form_item"
          hasFeedback
          validateStatus={formInfo.username.validationStatus}
          help={formInfo.username.help}
        >
          <Input
            data-testid="test_username_input"
            onBlur={e => formValidation(e)}
            prefix={<Icon type="user" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="username"
            type="text"
            name="fullName"
            value={user.fullName}
            onChange={event => handleChange(event)}
          />
        </Form.Item>
        <Form.Item
          data-testid="test_password_form_item"
          hasFeedback
          validateStatus={formInfo.password.validationStatus}
          help={formInfo.password.help}
        >
          <Input
            data-testid="test_password_input"
            onBlur={e => formValidation(e)}
            prefix={<Icon type="lock" style={{color: "rgba(0,0,0,.25)"}} />}
            placeholder="password"
            type="password"
            name="password"
            value={user.password}
            onChange={event => handleChange(event)}
          />
        </Form.Item>
        <Form.Item data-testid="test_submit_form_item">
          <Button
            data-testid="test_submit_button"
            type="primary"
            htmlType="submit"
            className={styles.registerFormButton}
            disabled={
              formInfo.email.validationStatus !== "success" ||
              formInfo.username.validationStatus !== "success" ||
              formInfo.password.validationStatus !== "success"
                ? true
                : false
            }
            loading={isLoading}
          >
            Register
          </Button>
          Or <Link to="/login">Login Here!</Link>
          <a className="google-auth" href={`${baseURL}/auth/google`}>
            <img
              className="google-icon"
              alt="google-icon"
              src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
            />
            SIGN UP WITH GOOGLE
          </a>
        </Form.Item>
      </Form>
      {error && (
        <Alert message={error} type="error" closable afterClose={() => setError(null)} />
      )}
    </div>
  );
}

const ConnectedForm = connect(null, {registerNewUser})(RegisterForm);

export default Form.create()(ConnectedForm);
