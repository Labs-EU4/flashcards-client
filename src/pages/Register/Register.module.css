.registerContainer {
  position: fixed;
  display: flex;
  flex-direction: column;
  width: 100%;
  align-items: center;
  margin: 30vh auto;
}

.registerForm {
  box-sizing: border-box;
  width: 100%;
  margin: 0 auto;
  margin-bottom: 2rem;
}

.registerFormButton {
  width: 100%;
}
