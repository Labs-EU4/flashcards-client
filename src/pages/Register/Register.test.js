import React from "react";
import {BrowserRouter} from "react-router-dom";
import * as rtl from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import {RegisterForm} from "./Register";
const mockRegister = jest.fn().mockResolvedValue(10);
let wrapper;
beforeEach(() => {
  wrapper = rtl.render(
    <BrowserRouter>
      <RegisterForm registerNewUser={mockRegister} />
    </BrowserRouter>
  );
});
const FormContainer = () => {
  return wrapper.queryByTestId("test_register_container");
};
const Form = () => {
  return wrapper.queryByTestId("test_register_form");
};
const EmailInput = () => {
  return wrapper.queryByTestId("test_email_input");
};
const UsernameInput = () => {
  return wrapper.queryByTestId("test_username_input");
};
const PasswordInput = () => {
  return wrapper.queryByTestId("test_password_input");
};
const confirmPasswordInput = () => {
  return wrapper.queryByTestId("test_confirmPassword_input");
};
const Button = () => {
  return wrapper.queryByTestId("test_submit_button");
};
const EmailWarning = () => {
  return wrapper.queryByText("Not a valid email");
};
const UsernameWarning = () => {
  return wrapper.queryByText("Username must be at least 5 characters");
};
const PasswordWarning = () => {
  return wrapper.queryByText("Password must be at least 5 characters");
};
const confirmPasswordWarning = () => {
  return wrapper.queryByText("Passwords do not match");
};
it("renders without crashing", () => {
  expect(wrapper.container).toMatchSnapshot();
});
it("renders all expected elements", () => {
  expect(FormContainer()).toBeInTheDocument();
  expect(Form()).toBeInTheDocument();
  expect(EmailInput()).toBeInTheDocument();
  expect(UsernameInput()).toBeInTheDocument();
  expect(PasswordInput()).toBeInTheDocument();
  expect(Button()).toBeInTheDocument();
});
it("does not submit due to submit button being disabled", () => {
  rtl.fireEvent.click(Button());
  expect(mockRegister).toHaveBeenCalledTimes(0);
});
it("does not make an axios call if some input fields are left blank", () => {
  rtl.fireEvent.submit(Form());
  expect(mockRegister).toHaveBeenCalledTimes(0);
});
it("Calls correct action on submit", () => {
  rtl.fireEvent.change(EmailInput(), {
    target: {value: "darragh@test.com"},
  });
  rtl.fireEvent.change(UsernameInput(), {
    target: {value: "Darragh Ferry"},
  });
  rtl.fireEvent.change(PasswordInput(), {
    target: {value: "123456789"},
  });
  rtl.fireEvent.change(confirmPasswordInput(), {
    target: {value: "123456789"},
  });
  rtl.fireEvent.submit(Form());
  expect(mockRegister).toHaveBeenCalledTimes(1);
  expect(mockRegister.mock.calls[0]).toMatchInlineSnapshot(`
    Array [
      Object {
        "email": "darragh@test.com",
        "fullName": "Darragh Ferry",
        "password": "123456789",
      },
    ]
  `);
});
it("Displays correct warning messages on input fields", () => {
  rtl.fireEvent.change(EmailInput(), {target: {value: "A"}});
  expect(EmailWarning()).toBeInTheDocument();
  rtl.fireEvent.change(UsernameInput(), {target: {value: "A"}});
  expect(UsernameWarning()).toBeInTheDocument();
  rtl.fireEvent.change(PasswordInput(), {target: {value: "A"}});
  expect(PasswordWarning()).toBeInTheDocument();
  rtl.fireEvent.change(confirmPasswordInput(), {target: {value: "B"}});
  expect(EmailWarning()).toBeInTheDocument();
});
afterEach(rtl.cleanup);
