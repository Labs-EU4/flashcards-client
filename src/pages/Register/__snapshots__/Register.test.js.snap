// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <div
    class="registerContainer"
    data-testid="test_register_container"
  >
    <form
      class="ant-form ant-form-horizontal registerForm"
      data-testid="test_register_form"
    >
      <div
        class="ant-row ant-form-item"
        data-testid="test_email_form_item"
      >
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <span
                class="ant-input-affix-wrapper"
              >
                <span
                  class="ant-input-prefix"
                >
                  <i
                    aria-label="icon: mail"
                    class="anticon anticon-mail"
                    style="color: rgba(0, 0, 0, 0.25);"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="mail"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0 0 68.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
                      />
                    </svg>
                  </i>
                </span>
                <input
                  class="ant-input"
                  data-testid="test_email_input"
                  name="email"
                  placeholder="email"
                  setfieldvalue=""
                  type="text"
                  value=""
                />
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
        data-testid="test_username_form_item"
      >
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <span
                class="ant-input-affix-wrapper"
              >
                <span
                  class="ant-input-prefix"
                >
                  <i
                    aria-label="icon: user"
                    class="anticon anticon-user"
                    style="color: rgba(0, 0, 0, 0.25);"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="user"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M858.5 763.6a374 374 0 0 0-80.6-119.5 375.63 375.63 0 0 0-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 0 0-80.6 119.5A371.7 371.7 0 0 0 136 901.8a8 8 0 0 0 8 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 0 0 8-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                      />
                    </svg>
                  </i>
                </span>
                <input
                  class="ant-input"
                  data-testid="test_username_input"
                  name="fullName"
                  placeholder="username"
                  setfieldvalue=""
                  type="text"
                  value=""
                />
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
        data-testid="test_password_form_item"
      >
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <span
                class="ant-input-affix-wrapper"
              >
                <span
                  class="ant-input-prefix"
                >
                  <i
                    aria-label="icon: lock"
                    class="anticon anticon-lock"
                    style="color: rgba(0, 0, 0, 0.25);"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="lock"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M832 464h-68V240c0-70.7-57.3-128-128-128H388c-70.7 0-128 57.3-128 128v224h-68c-17.7 0-32 14.3-32 32v384c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V496c0-17.7-14.3-32-32-32zM332 240c0-30.9 25.1-56 56-56h248c30.9 0 56 25.1 56 56v224H332V240zm460 600H232V536h560v304zM484 701v53c0 4.4 3.6 8 8 8h40c4.4 0 8-3.6 8-8v-53a48.01 48.01 0 1 0-56 0z"
                      />
                    </svg>
                  </i>
                </span>
                <input
                  class="ant-input"
                  data-testid="test_password_input"
                  name="password"
                  placeholder="password"
                  setfieldvalue=""
                  type="password"
                  value=""
                />
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
        data-testid="test_submit_form_item"
      >
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control"
          >
            <span
              class="ant-form-item-children"
            >
              <button
                class="ant-btn registerFormButton ant-btn-primary"
                data-testid="test_submit_button"
                type="submit"
              >
                <span>
                  Register
                </span>
              </button>
              Or 
              <a
                href="/login"
              >
                Login Here!
              </a>
            </span>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
